src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs

```diff
@@ -207,53 +231,131 @@ internal static unsafe byte[] UnsafeDecode(ReadOnlySpan<char> strSpan)
+            // If the incoming chars don't contain any of the base64url characters that need to be replaced,
+            // and if the incoming chars are of the exact right length, then we'll be able to just pass the
+            // incoming chars directly to DecodeFromUtf8InPlace. Otherwise, rent an array, copy all the
+            // data into it, and do whatever fixups are necessary on that copy, then pass that copy into
+            // DecodeFromUtf8InPlace.
+
+            const int StackAllocThreshold = 512;
+            char[] arrayPoolChars = null;
+            scoped Span<char> charsSpan = default;
+            scoped ReadOnlySpan<char> source = strSpan;
+
+            if (needReplace || decodedLength != source.Length)
+            {
+                charsSpan = decodedLength <= StackAllocThreshold ?
+                    stackalloc char[StackAllocThreshold] :
+                    arrayPoolChars = ArrayPool<char>.Shared.Rent(decodedLength);
+                charsSpan = charsSpan.Slice(0, decodedLength);
+
+                source = HandlePaddingAndReplace(charsSpan, source, needReplace);
             }
 
             byte[] arrayPoolBytes = null;
             Span<byte> bytesSpan = decodedLength <= StackAllocThreshold ?
                 stackalloc byte[StackAllocThreshold] :
                 arrayPoolBytes = ArrayPool<byte>.Shared.Rent(decodedLength);
 
-            bool converted = Convert.TryFromBase64Chars(source, bytesSpan, out int bytesWritten);
-            Debug.Assert(converted, "Expected TryFromBase64Chars to be successful");
-            byte[] result = bytesSpan.Slice(0, bytesWritten).ToArray();
+            int length = Encoding.UTF8.GetBytes(source, bytesSpan);
+            Span<byte> utf8Span = bytesSpan.Slice(0, length);
 
-            if (arrayPoolBytes is not null)
+            try
             {
-                bytesSpan.Clear();
-                ArrayPool<byte>.Shared.Return(arrayPoolBytes);
+                OperationStatus status = System.Buffers.Text.Base64.DecodeFromUtf8InPlace(utf8Span, out int bytesWritten);
+                if (status != OperationStatus.Done)
+                    throw LogHelper.LogExceptionMessage(new FormatException(LogHelper.FormatInvariant(LogMessages.IDX10400, strSpan.ToString())));
+
+                utf8Span.Slice(0, bytesWritten).CopyTo(output);
             }
+            finally
+            {
+                if (arrayPoolBytes is not null)
+                {
+                    bytesSpan.Clear();
+                    ArrayPool<byte>.Shared.Return(arrayPoolBytes);
+                }
 
-            if (arrayPoolChars is not null)
+                if (arrayPoolChars is not null)
+                {
+                    charsSpan.Clear();
+                    ArrayPool<char>.Shared.Return(arrayPoolChars);
+                }
+            }
+        }
+
+        private static ReadOnlySpan<char> HandlePaddingAndReplace(Span<char> charsSpan, ReadOnlySpan<char> source, bool needReplace)
```

Comment:
nit: Rename `charsSpan` to something like `destination` since we're copying into it.
Reorder params so that source comes first, then charsSpan/destination. Easier to read that way source > destination.
Since we're updating the contents of `charsSpan`, we don't need to return it? Right now it looks like we're returning a brand new Span.

Reply:
Ignore point 3, looks like it matches the previous behavior.